### PR TITLE
Fixes for using dblib PDO driver.

### PR DESCRIPTION
**Pull request by ajoneil**
These fixes allow *nix environments to connect to SQL Server using
the dblib PDO driver and the silverstripe mssql module.

  - Only set MYSQL_ATTR_INIT_COMMAND when using the mysql driver, this
    constant isn't defined if the mysql pdo driver isn't installed
  - Supress warnings on getting the server version, attributes aren't
    supported by the dblib driver
  - Explicitly check for errors on sql exec, checking the return
    value isn't reliable for statements with no return value (e.g.
    USE database)

### CODE REVIEW

File: model/connect/PDOConnector.php [hunk 1]
```diff
@@ -227,7 +228,7 @@ public function exec($sql, $errorLevel = E_USER_ERROR) {
 		$result = $this->pdoConnection->exec($sql);
 
 		// Check for errors
-		if ($result !== false) {
+		if (!$this->hasError($result)) {
```
Comment (tractorcow) [inline model/connect/PDOConnector.php]: The problem is that `false` as a return value now reports hasError() = false, when before it was false. Docs on this method:

```
PDO::exec() returns the number of rows that were modified or deleted by the SQL statement you issued. If no rows were affected, PDO::exec() returns 0.
```

I think the `!== false` check was correct.

Comment (tractorcow) [inline model/connect/PDOConnector.php]: You could also fix it by making hasError() return true for `=== false`

Comment (ajoneil) [inline model/connect/PDOConnector.php]: I'm not sure what you mean by the `false` return value here - this still returns the result set if the query was successful, or `null` if there was an error.

The `PDO::exec()` docs don't match with what I'm seeing - this is returning `false` with a `USE database` statement with the dblib driver, but the database select is successful.

Comment (ajoneil) [inline model/connect/PDOConnector.php]: I reorganised the code a little here, so it's easier to read.

**Comment by dhensby** [comment]
Although this is something we can only do in 4, really anything we are finding we have to wrap in "is this a xxx kind of driver" should be logic that each connector provides.

Each connector should provide the connection options, the DSN, and so on.

File: model/connect/PDOConnector.php [hunk 2]
```diff
@@ -178,7 +179,7 @@ public function connect($parameters, $selectDB = false) {
 	}
 
 	public function getVersion() {
-		return $this->pdoConnection->getAttribute(PDO::ATTR_SERVER_VERSION);
+		return @$this->pdoConnection->getAttribute(PDO::ATTR_SERVER_VERSION);
```
Comment (dhensby) [inline model/connect/PDOConnector.php]: Why do we need to suppress errors here?

Comment (ajoneil) [inline model/connect/PDOConnector.php]: "Supress warnings on getting the server version, attributes aren't supported by the dblib driver"

**Comment by tractorcow** [comment]
Damn it, I wrote some comments 6 days ago but forgot to submit review (so they were hidden). Sorry!

**Comment by tractorcow** [comment]
I think we need to use `hasError($this->pdoConnection)` instead of `hasError($result)` here.

**Comment by ajoneil** [comment]
Looks like you're right with checking the error on the connection, I'll fix that up.

**Comment by chillu** [comment]
Yay a wild Andrew appears!

**Comment by tractorcow** [comment]
Looks good; Let's see if travis wants to work today.

**Comment by sminnee** [comment]
@tractorcow are you happy with the latest incarnation of this?

**Comment by tractorcow** [comment]
Yes. :)

**Pull request merged**